src/main/scala/tech/cryptonomic/conseil/tezos/ApiOperations.scala

```diff
@@ -643,6 +643,12 @@ object ApiOperations {
     Map("block" -> block, "operation_groups" -> operationGroups)
   }
 
+  def blockExists(hash: String)(implicit ec: ExecutionContext): Future[Boolean] =
```

Comment:
Is this an APIOperation, so that the user can check if a specific hash exists in the chain, or a utility function to help with testing? Also, this would need logic to account for forks

Reply:
Good point, it's a utility so it would be fine to make it `private`.

I used this because of this code in the original `processBlocks`

```scala
  else if(block.metadata.header.level <= minLevel && followFork)
    ApiOperations.fetchBlock(block.metadata.header.predecessor) match {
      case Success(_)  => block :: blockSoFar
      case Failure(_)    => processBlocks(network, block.metadata.header.predecessor, minLevel, maxLevel, block :: blockSoFar, followFork)
          }
```
where it loads the predecessor block, seemingly for no other reason than to check that it's already in the db.

The corresponding excerpt in the rewritten code is
```scala
  else if (belowMinWithForking)
    ApiOperations.blockExists(current.header.predecessor) flatMap { predecessorStored =>
      if (predecessorStored)
        Future.successful(current :: collected) //pred on db, we're done here
      else
        getMetadata(network, current.header.predecessor) flatMap { pred =>
          recursiveMetadataFetch(network, pred, minLevel, maxLevel, followFork, current :: collected)
        }
      }
```

I would've simplified further to only check for `Block` existence and not for the associated `OperationGroups`, but it would've been more than a refactor, as the original code expects both to be available.